aptos-move/framework/aptos-framework/sources/delegation_pool.move

```diff
@@ -1018,6 +1072,100 @@ module aptos_framework::delegation_pool {
+                DelegatorLock[delegator_addr].pools.upsert(pool_address, proposal_expiration);
+            };
+        };
+    }
+
+    /// Allows a voter to vote on proposals when the pool's lockup is less than the required lockup to vote on the
+    /// specified proposal. To call this function, the voter needs to specify all delegators that have delegated the
+    /// votes to them (including themselves if they have shares in the delegation pool). This is necessary to check
+    /// that all the delegators have extra lockups that are sufficient to cover the voting duration of the governance.
+    ///
+    /// For simplicity, no partial voting is allowed. This will vote with all remaining voting power.
+    public entry fun vote_with_active_locks(
+        voter: &signer,
+        pool_address: address,
+        delegators: vector<address>,
+        proposal_id: u64,
+        should_pass: bool
+    ) acquires BeneficiaryForOperator, DelegatorLock, DelegationPool, GovernanceRecords, NextCommissionPercentage {
+        let voter_addr = signer::address_of(voter);
+        let total_delegator_active_shares = 0;
+        let total_delegator_pending_inactive_shares = 0;
+        let proposal_expiration = aptos_governance::get_proposal_expiration(proposal_id);
+        delegators.for_each(|delegator| {
+            // If any delegator delegates to someone else or the voter will be updated soon, disallow voting to avoid
+            // double voting.
+            let (voter, pending_voter, _) = calculate_and_update_voting_delegation(pool_address, delegator);
+            assert!(
+                voter == voter_addr && (pending_voter == @0x0 || pending_voter == voter_addr),
+                EONE_OR_MORE_DELEGATIONS_MISMATCH
+            );
+
+            // Make sure either the pool's lockup is sufficient or the delegator has an extra lock longer than the
+            // voting duration of the proposal.
+            let lock_expiration = delegator_lockup_expiration(delegator, pool_address);
+            assert!(lock_expiration >= proposal_expiration, EINSUFFICIENT_LOCKUP);
+
+            // Need to borrow this in every iteration instead of outside to avoid multiple references from function
+            // calls above this line.
+            let delegation_pool = &DelegationPool[pool_address];
+            total_delegator_active_shares += get_delegator_active_shares(delegation_pool, delegator);
+            total_delegator_pending_inactive_shares += get_delegator_pending_inactive_shares(delegation_pool, delegator);
+        });
+
+        let governance_records = &mut GovernanceRecords[pool_address];
+        let delegation_pool = &DelegationPool[pool_address];
+        let delegated_votes =
+            update_and_borrow_mut_delegated_votes(delegation_pool, governance_records, voter_addr);
+        // All specified delegators must account for 100% of the voter's delegated votes. This prevents voters from
```

Comment:
why?
so it is required that all delegator delegating to this voter must call `lock_shares` otherwise, the voter practically cannot vote? This requirement is too strict I think.

Reply:
Yes that's correct. Staking and delegation contracts are designed to allow a separate voter address from the staker or delegator address. This means the voter might not actually own the underlying assets. It's risky to allow voter to add an extra lockup as that'd make the APT delegation stuck and affect the actual delegator (owner of funds). I don't yet see any other way to make this work without this unfortunately more complex process